pyscriptjs/tests/integration/support.py

```diff
@@ -105,47 +105,78 @@ def init_page(self, page):
         # page and they should not cause the test to fail, you should call
-        # self.check_errors() in the test itself.
-        self.check_errors()
+        # self.check_js_errors() in the test itself.
+        self.check_js_errors()
+
+    def _on_console(self, msg):
+        self.console.add_message(msg.type, msg.text)
 
     def _on_pageerror(self, error):
-        self.logger.log("JS exception", error.stack, color="red")
-        self._page_errors.append(error)
+        self.console.add_message("js_error", error.stack)
+        self._js_errors.append(error)
 
-    def check_errors(self):
+    def check_js_errors(self, *expected_messages):
         """
         Check whether JS errors were reported.
 
-        If it finds a single JS error, raise JsError.
-        If it finds multiple JS errors, raise JsMultipleErrors.
+        expected_messages is a list of strings of errors that you expect they
+        were raised in the page.  They are checked using a simple 'in' check,
+        equivalent to this:
+            if expected_message in actual_error_message:
+                ...
+
+        If an error was expected but not found, it raises
+        DidNotRaiseJsError().
+
+        If there are MORE errors other than the expected ones, it raises an
+        exception:
+
+          - if it finds a single JS error, raise JsError
+          - if it finds multiple JS errors, raise JsMultipleErrors
 
         Upon return, all the errors are cleared, so a subsequent call to
-        check_errors will not raise, unless NEW JS errors have been reported
+        check_js_errors will not raise, unless NEW JS errors have been reported
         in the meantime.
         """
-        exc = None
-        if len(self._page_errors) == 1:
-            # if there is a single error, wrap it
-            exc = JsError(self._page_errors[0])
-        elif len(self._page_errors) >= 2:
-            exc = JsMultipleErrors(self._page_errors)
-        self._page_errors = []
-        if exc:
-            raise exc
-
-    def clear_errors(self):
+        expected_messages = list(expected_messages)
+        js_errors = self._js_errors[:]
+
+        for i, msg in enumerate(expected_messages):
+            for j, error in enumerate(js_errors):
+                if msg is not None and error is not None and msg in error.message:
+                    # we matched one expected message with an error, remove both
+                    expected_messages[i] = None
+                    js_errors[j] = None
+
+        # if everything is find, now expected_messages and js_errors contains
+        # only Nones. If they contain non-None elements, it means that we
+        # either have messages which are expected-but-not-found or errors
+        # which are found-but-not-expected.
+        expected_messages = [msg for msg in expected_messages if msg is not None]
+        js_errors = [err for err in js_errors if err is not None]
+        self.clear_js_errors()
```

Comment:
[question] Why are we cleaning it here?

Reply:
because `self.check_js_errors()` is also automatically called at teardown, after executing each test:
https://github.com/pyscript/pyscript/blob/3a8a64e49158a5645cfab6b875821a321ee1c58c/pyscriptjs/tests/integration/support.py#L112-L117

if we don't clear the errors here, you would get the same exception twice (once when you call `check_js_errors()` explicitly, and once at teardown)